### ISSUE TITLE
Datalist returns null from block settings

### ISSUE THREAD

**Issue by bjarnef** [leekelleher/umbraco-contentment#111]
### What is the current behaviour?

We have a custom data source to select some devices / screen resolutions to hide some block elements, where we use some block settings.

```
public class DeviceDataSource : IDataListSource, IDataListSourceValueConverter
{
    private readonly Dictionary<string, DeviceType> _lookup;

    public DeviceDataSource()
    {
        var devices = new List<DeviceType>
        {
            new DeviceType {
                Name = "Mobile",
                Key = "mobile",
                Icon = "icon-iphone",
            },
            new DeviceType {
                Name = "Tablet",
                Key = "tablet",
                Icon = "icon-ipad",
            },
            new DeviceType {
                Name = "Desktop",
                Key = "desktop",
                Icon = "icon-display",
            },
        };
        _lookup = devices.ToDictionary(x => x.Key.ToString());
    }

    public string Name => "Devices";

    public string Description => "Data source for devices.";

    public string Icon => "icon-brick";

    public OverlaySize OverlaySize => OverlaySize.Small;

    public Dictionary<string, object> DefaultValues => default;

    public IEnumerable<ConfigurationField> Fields => default;

    public IEnumerable<DataListItem> GetItems(Dictionary<string, object> config)
    {
        return _lookup
            .Select(x => new DataListItem
            {
                Name = x.Value.Name,
                Value = x.Key,
                Icon = x.Value.Icon,
                Description = ""
            });
    }

    public Type GetValueType(Dictionary<string, object> config) => typeof(DeviceType);

    public object ConvertValue(Type type, string value) => _lookup.ContainsKey(value) == true ? _lookup[value] : null;
}

public class DeviceType
{
    public string Name { get; set; }
    public string Key { get; set; }
    public string Icon { get; set; }
}
```

In this use-case the content types also vary by culture, but not the element types or block settings.

This works fine when using the property directly on a Settings node (which vary by culture).
However if accessing the property from a Settings element type on a block element placed in Block List on a Standard Page (which vary by culture) the property returns null instead of the selected items.

### What is the expected behaviour?

The property on either block content or block settings to return the selected items.

### Please tell us about your set-up:

- Contentment version: [1.4.2]
- Umbraco version: [8.9.0]
- Umbraco ModelsBuilder is: [enabled]
- Browser: [all]

### Anything else?

**Settings node**

![image](https://user-images.githubusercontent.com/2919859/119131048-434db380-ba39-11eb-88ea-2f02167b9507.png)

**Block element settings**

![image](https://user-images.githubusercontent.com/2919859/119133147-d851ac00-ba3b-11eb-8dde-aa5e7d2b19c2.png)

**Block element**

![image](https://user-images.githubusercontent.com/2919859/119131156-6710f980-ba39-11eb-8a29-27ace3c3508d.png)

**Comment by bjarnef** [comment]
It does however work correct with these settings based on enum values, so I wonder if it is specific related to a custom data source?

![image](https://user-images.githubusercontent.com/2919859/119132278-ccb1b580-ba3a-11eb-8ad4-f9ea8825603d.png)

**Comment by leekelleher** [comment]
@bjarnef I haven't looked into the issue with the Block List Settings yet, but wanted to pick up on the use of the [`EnumMember` attribute](https://docs.microsoft.com/en-us/dotnet/api/system.runtime.serialization.enummemberattribute?view=netframework-4.7.2) - that the Enum data-source does not take it into account, so the value stored is the Enum's name.

I wasn't even aware of the `EnumMember` attribute before seeing your example above. 😮 

At present, the Enum data-source is only aware of the [`DataListItemAttribute`](https://github.com/leekelleher/umbraco-contentment/blob/1.4.2/src/Umbraco.Community.Contentment/DataEditors/DataList/DataListItemAttribute.cs) and [`DescriptionAttribute`](https://docs.microsoft.com/en-us/dotnet/api/system.componentmodel.descriptionattribute?view=netframework-4.7.2).
<https://github.com/leekelleher/umbraco-contentment/blob/1.4.2/src/Umbraco.Community.Contentment/DataEditors/DataList/DataSources/EnumDataListSource.cs#L80-L86>

---

re: Block List, I haven't actually used it on any projects myself yet. I'll get around to looking at it.

**Comment by bjarnef** [comment]
@leekelleher ahh okay, not a bit issue though as I compare with the enum and not specifically the value in `EnumMember`. However if re-using it in other part of code, `EnumMember` is useful for serialization of data.

**Comment by bjarnef** [comment]
@leekelleher I am sure you are busy with many other projects and v2.0 of Contentment 😎 
I think it may be something inside the `DataListValueConverter`.
https://github.com/leekelleher/umbraco-contentment/blob/develop/src/Umbraco.Community.Contentment/DataEditors/DataList/DataListValueConverter.cs

**Comment by leekelleher** [comment]
@bjarnef I've tried taking a look into this over the past hour or so... I can reproduce the issue, but it looks to be something how Umbraco is calling the ValueConverter methods.  This issue goes beyond Contentment/DataList.

Here's where I got to with my debugging...

On a Block List item, when accessing a `.Content`'s property value, it will call the [`ConvertSourceToIntermediate()` method](https://github.com/leekelleher/umbraco-contentment/blob/1.4.2/src/Umbraco.Community.Contentment/DataEditors/DataList/DataListValueConverter.cs#L37) first, then it will call the [`ConvertIntermediateToObject()` method](https://github.com/leekelleher/umbraco-contentment/blob/1.4.2/src/Umbraco.Community.Contentment/DataEditors/DataList/DataListValueConverter.cs#L52). However when accessing a `.Settings`' property value, it will only call the [`ConvertIntermediateToObject()` method](https://github.com/leekelleher/umbraco-contentment/blob/1.4.2/src/Umbraco.Community.Contentment/DataEditors/DataList/DataListValueConverter.cs#L52) - it does not call `ConvertSourceToIntermediate()`. 😢 

This means that the `inter` value being passed in to DataList's ValueConverter is in the `source` format, in this particular case, a `JArray`, so the `ConvertSourceToIntermediate()` doesn't know what to do with it.  _(A workaround would be for it to support `JArray` objects too. But that's a workaround, not a fix to the underlying issue.)_

I started to dig around Umbraco's source, to see what I could find. I got as far as here: [`PublishedElementPropertyBase`'s `GetInterValue()` method](https://github.com/umbraco/Umbraco-CMS/blob/release-8.9.0/src/Umbraco.Web/PublishedCache/PublishedElementPropertyBase.cs#L153), this does have an initialized flag, so I'm thinking that somehow that's already being set (somehow? 🤷) when the `.Settings` element is being used/called.

At this point, I don't have much time to start debugging Umbraco core, _(sorry)._ All I know is that would impact other property-editor value-converters, not only Contentment/DataList. 😬

**Comment by bjarnef** [comment]
@leekelleher thanks for looking into this -  really appreciated 😊🙌
Okay, so if sound a workaround could be to add the property as Content on the block element type instead of Settings element.

Maybe @warrenbuckley has some deeper insights how the value converter works regarding the Block List and in this case `Settings` element?

**Comment by leekelleher** [comment]
@bjarnef I've put a workaround fix in commit e03cb2114e8691f06af359a9b06017bc8197002a, currently in the v2.0 branch. I'll look to backport it for a v1.4.3 patch release.

**Comment by bjarnef** [comment]
@leekelleher cool, would be a great workaround for now. Would be great to investigate further how the `.Settings` could return the correct type though, but this workaround might be good enough for now 😁

**Comment by leekelleher** [comment]
@bjarnef FYI, released [v1.4.3](https://github.com/leekelleher/umbraco-contentment/releases/tag/1.4.3).
